nototools/generate_coverage_data.py

```diff
@@ -0,0 +1,253 @@
+# limitations under the License.
+
+import argparse
+import collections
+import datetime
+from os import path
+import sys
+
+from nototools import cmap_data
+from nototools import coverage
+from nototools import tool_utils
+
+from xml.etree import ElementTree as ET
+
+"""Generate a coverage.xml file listing the codepoints covered by
+a font family plus a name."""
+
+default_coverage_file = 'noto_cmap_phase2.xml'
+
+MetaData = collections.namedtuple('MetaData', 'date,program,args')
+CmapData = collections.namedtuple('CmapData', 'name,ranges')
+CoverageData = collections.namedtuple('CoverageData', 'metadata,cmapdata')
+
+
+def get_cps_from_files(paths):
+  """Return a tuple of the cps and the paths that we actually read."""
+  cps = set()
+  new_paths = set()
+  for f in paths:
+    ext = path.splitext(f)[1]
+    if ext not in ['.otf', '.ttf']:
+      continue
+    cps |= coverage.character_set(f)
+    new_paths.add(f)
+  return cps, sorted(new_paths)
+
+
+def get_cps_from_cmap_data_file(data_file):
+  cps = set()
+  data = cmap_data.read_cmap_data_file(data_file)
+  for row in data.table.rows:
+    cps |= tool_utils.parse_int_ranges(row.ranges)
+  return cps
+
+
+def _create_metadata(**kwargs):
+  """Create a MetaData object from the args.  'date' defaults to today's
+  date."""
+  date = str(kwargs.pop('date', datetime.date.today()))
+  program = str(kwargs.pop('program', 'generate_coverage_data'))
+  arglist = [
+      (k, v) for k, v in sorted(kwargs.iteritems())
+      if v is not None]
+  return MetaData(date, program, arglist)
+
+
+def _create_cmapdata(name, cmap):
+  """Create a CmapData object from the name and cmap."""
+  return CmapData(name, tool_utils.write_int_ranges(cmap))
+
+
+def create(name, cps, paths=None, cmap_data=None):
+  """Generate the coverage data object."""
+  metadata = _create_metadata(paths=paths, cmap_data=cmap_data)
+  cmapdata = _create_cmapdata(name, cps)
+  return CoverageData(metadata, cmapdata)
+
+
+def _common_path_prefix(items):
+  """Assuming items is an array of paths using '/' as a path separator, return a
+  common path prefix of the items."""
+  prefix = None
+  if len(items) <= 1:
+    return ''
+  for item in items:
+    if prefix is None:
+      last = item.rfind('/')
+      if last < 0:
+        return ''
+      prefix = item[:last + 1]
+      continue
+    end = len(item)
+    while True:
+      last = item.rfind('/', 0, end)
+      if last < 0:
+        return ''
+      item_prefix = item[:last + 1]
+      if prefix == item_prefix:
+        break
+      if prefix.startswith(item_prefix):
+        prefix = item_prefix
+        break
+      end = last
+  return prefix
+
+
+def _build_meta_elem(metadata):
+  """Convert metadata to ET node."""
+  meta = ET.Element('meta', date=metadata.date, program=metadata.program)
+  if metadata.args:
+    args = ET.Element('args')
+    for k, v in metadata.args:
+      if isinstance(v, list):
+        arg = ET.Element(k, {'isList': 'true'})
```

Comment:
Is there any difference between providing attributes as keyword args vs. a dictionary? (if not, why use both)